base/parse.jl

```diff
@@ -194,3 +194,6 @@ function parse(str::AbstractString; raise::Bool=true)
     end
     return ex
 end
+
+promote_op{R<:AbstractFloat,S<:AbstractString}(::typeof(parse), ::Type{R}, ::Type{S}) = R
```

Comment:
Yeah, but this still looks like return type declarations, e.g.:
```julia
parse{T}(::Type{T}, x)::T = ...
```

That said, maybe that's the best short-term approach. I just wanted to note that longer-term, we shouldn't have two parallel systems to compute return types.